src/main/java/ch/njol/skript/events/EvtGrow.java

```diff
@@ -34,56 +39,188 @@ public class EvtGrow extends SkriptEvent {
+				.examples(
+					"on grow:",
+					"on grow of tree:",
+					"on grow of wheat[age=7]:",
+					"on grow from a sapling:",
+					"on grow into tree:",
+					"on grow from a sapling into tree:",
+					"on grow of wheat, carrots, or potatoes:",
+					"on grow into tree, giant mushroom, cactus:",
+					"on grow from wheat[age=0] to wheat[age=1] or wheat[age=2]:")
+				.since("1.0, 2.2-dev20 (plants), INSERT VERSION (from, into, blockdata)");
 	}
 	
 	@Nullable
-	private Literal<StructureType> types;
+	private Literal<Object> toTypes;
 	@Nullable
-	private Literal<ItemType> blocks;
-	private int eventType;
+	private Literal<Object> fromTypes;
+
+	// Restriction on the type of grow event, ANY, STRUCTURE or BLOCK
+	private int eventRestriction;
+
+	// Restriction on the type of action, OF, FROM, INTO, or FROM_INTO
+	private int actionRestriction;
 	
 	@SuppressWarnings("unchecked")
 	@Override
 	public boolean init(Literal<?>[] args, int matchedPattern, ParseResult parseResult) {
-		eventType = parseResult.mark; // ANY, STRUCTURE or BLOCK
-		if (eventType == STRUCTURE) {
-			types = (Literal<StructureType>) args[0];
-		} else if (eventType == BLOCK) {
-			blocks = (Literal<ItemType>) args[1]; // Arg 1 may not be present... but it is in the array still, as null
+		eventRestriction = parseResult.mark; // ANY, STRUCTURE or BLOCK
+		actionRestriction = matchedPattern; // OF, FROM, INTO, or FROM_INTO
+
+		switch (actionRestriction) {
+			case OF:
+				if (eventRestriction == STRUCTURE) {
+					fromTypes = (Literal<Object>) args[0];
+				} else if (eventRestriction == BLOCK) {
+					fromTypes = (Literal<Object>) args[1];
+				}
+				break;
+			case FROM:
+				fromTypes = (Literal<Object>) args[0];
+				break;
+			case INTO:
+				if (eventRestriction == STRUCTURE) {
+					toTypes = (Literal<Object>) args[0];
+				} else if (eventRestriction == BLOCK) {
+					toTypes = (Literal<Object>) args[1];
+				}
+				break;
+			case FROM_INTO:
+				fromTypes = (Literal<Object>) args[0];
+				if (eventRestriction == STRUCTURE) {
+					toTypes = (Literal<Object>) args[1];
+				} else if (eventRestriction == BLOCK) {
+					toTypes = (Literal<Object>) args[2];
+				}
+				break;
+			default:
+				assert false;
+				return false;
 		}
-		// Else: no type restrictions specified
 		return true;
 	}
 	
 	@Override
 	public boolean check(Event event) {
-		if (eventType == STRUCTURE && types != null && event instanceof StructureGrowEvent) {
-			return types.check(event, type -> type.is(((StructureGrowEvent) event).getSpecies()));
-		} else if (eventType == BLOCK && blocks != null && event instanceof BlockGrowEvent) {
-			return blocks.check(event, type -> type.isOfType(((BlockGrowEvent) event).getNewState()));
+		if (fromTypes == null && actionRestriction != INTO)
+			// True for "on grow", false for anything else.
+			return actionRestriction == OF;
```

Comment:
Maybe add comment here why it's `== OF`. (matchedPattern)

Reply:
I thought that's what the comment that's already there explains? That it makes it return true for `on grow` and false for anything else? Is it not clear enough?